We will be looking at a dataset of used cars from eBay Kleinanzeigen, a classifieds section of the German eBay website. This dataset contains 50,000 slightly modified data points, which is actually a sample of the original dataset originally scraped and uploaded to Kaggle. We will be cleaning the data and analyzing the included used car listing.

# Opening the File, Formatting

In [410]:
# Import the libraries we need
import numpy as np
import pandas as pd

In [411]:
# Read the autos.csv file into pandas
autos = pd.read_csv("autos.csv", encoding = 'Latin-1')

# Could not read the file using the default encoding, 'UTF-8'
# without error

In [413]:
autos.info()
autos.head()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
dateCrawled            50000 non-null object
name                   50000 non-null object
seller                 50000 non-null object
offerType              50000 non-null object
price                  50000 non-null object
abtest                 50000 non-null object
vehicleType            44905 non-null object
yearOfRegistration     50000 non-null int64
gearbox                47320 non-null object
powerPS                50000 non-null int64
model                  47242 non-null object
odometer               50000 non-null object
monthOfRegistration    50000 non-null int64
fuelType               45518 non-null object
brand                  50000 non-null object
notRepairedDamage      40171 non-null object
dateCreated            50000 non-null object
nrOfPictures           50000 non-null int64
postalCode             50000 non-null int64
lastSeen               50000 non-null obj

dateCrawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offerType   price   abtest vehicleType  yearOfRegistration  \
0  privat   Angebot  $5,000  control         bus                2004   
1  privat   Angebot  $8,500  control   limousine                1997   
2  privat   Angebot  $8,990     test   limousine                2009   
3  privat   Angebot  $4,350  control  kleinwagen                2007   
4  privat   Angebot  $1,350     test       kombi                2003   

     gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0    manuell      158  andere  150,000km                    3      lpg   
1  automatik      286     7er  150,000km                    6   benzin   
2    manuell      102    golf   70,000km                    7   benzin   
3  automatik       71  fortwo   70,000km                    6   benzin   
4    manuell        0   focus  150,000km                    7   benzin   

        brand notRepairedDamage          dateCreated  nrOfPictures  \
0     peugeot              nein  2016-03-26 00:00:00             0   
1         bmw              nein  2016-04-04 00:00:00             0   
2  volkswagen              nein  2016-03-26 00:00:00             0   
3       smart              nein  2016-03-12 00:00:00             0   
4        ford              nein  2016-04-01 00:00:00             0   

   postalCode             lastSeen  
0       79588  2016-04-06 06:45:54  
1       71034  2016-04-06 14:45:08  
2       35394  2016-04-06 20:15:37  
3       33729  2016-03-15 03:16:28  
4       39218  2016-04-01 14:38:50

The autos dataset does contain 50,000 data points. There are 20 columns, most of which are strings. Null values are expected in the following columns:
- vehicleType
- gearbox
- model
- fuelType
- notRepairedDamage

None of these columns have more than ~20% null values.

The column names also use camelcase instead of snakecase (i.e. no underscores between words).

In [414]:
autos.columns

Index(['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'yearOfRegistration', 'gearbox', 'powerPS', 'model',
       'odometer', 'monthOfRegistration', 'fuelType', 'brand',
       'notRepairedDamage', 'dateCreated', 'nrOfPictures', 'postalCode',
       'lastSeen'],
      dtype='object')

In [415]:
autos.columns = ['date_crawled', 'name', 'seller', 'offer_type', 'price', 'abtest',
       'vehicle_type', 'registration_year', 'gearbox', 'power_PS', 'model',
       'odometer', 'registration_month', 'fuel_type', 'brand',
       'unrepaired_damage', 'ad_created', 'nr_of_pictures', 'postal_code',
       'last_seen']

autos.head()

date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offer_type   price   abtest vehicle_type  registration_year  \
0  privat    Angebot  $5,000  control          bus               2004   
1  privat    Angebot  $8,500  control    limousine               1997   
2  privat    Angebot  $8,990     test    limousine               2009   
3  privat    Angebot  $4,350  control   kleinwagen               2007   
4  privat    Angebot  $1,350     test        kombi               2003   

     gearbox  power_PS   model   odometer  registration_month fuel_type  \
0    manuell       158  andere  150,000km                   3       lpg   
1  automatik       286     7er  150,000km                   6    benzin   
2    manuell       102    golf   70,000km                   7    benzin   
3  automatik        71  fortwo   70,000km                   6    benzin   
4    manuell         0   focus  150,000km                   7    benzin   

        brand unrepaired_damage           ad_created  nr_of_pictures  \
0     peugeot              nein  2016-03-26 00:00:00               0   
1         bmw              nein  2016-04-04 00:00:00               0   
2  volkswagen              nein  2016-03-26 00:00:00               0   
3       smart              nein  2016-03-12 00:00:00               0   
4        ford              nein  2016-04-01 00:00:00               0   

   postal_code            last_seen  
0        79588  2016-04-06 06:45:54  
1        71034  2016-04-06 14:45:08  
2        35394  2016-04-06 20:15:37  
3        33729  2016-03-15 03:16:28  
4        39218  2016-04-01 14:38:50

The column names havebeen modified from camelcase to snakecase. This ensures that the column names are more readable while still counting as one string element each. This can be useful when columns need to be called from this DataFrame.

In [416]:
autos.describe()

registration_year      power_PS  registration_month  nr_of_pictures  \
count       50000.000000  50000.000000        50000.000000         50000.0   
mean         2005.073280    116.355920            5.723360             0.0   
std           105.712813    209.216627            3.711984             0.0   
min          1000.000000      0.000000            0.000000             0.0   
25%          1999.000000     70.000000            3.000000             0.0   
50%          2003.000000    105.000000            6.000000             0.0   
75%          2008.000000    150.000000            9.000000             0.0   
max          9999.000000  17700.000000           12.000000             0.0   

        postal_code  
count  50000.000000  
mean   50813.627300  
std    25779.747957  
min     1067.000000  
25%    30451.000000  
50%    49577.000000  
75%    71540.000000  
max    99998.000000

None of the cars listed had pictures attached to their ads. We can drop the nr_of_pictures column as the values are the same, meaning there is no useful information for analysis. The same can be said for the seller and offer_Type columns.

# Filtering Outliers

Registration year, power_PS and registration month are worth investigating. There are additional columns that contain numeric data that is stored as text and should be cleaned, such as price and odometer. 

In [417]:
autos['price'] = (autos['price'].str.replace('$','')
                                .str.replace(',','')
                                .astype(float))
autos['odometer'] = (autos['odometer'].str.replace('km','')
                                      .str.replace(',','')
                                      .astype(float))
autos.rename(columns = {'odometer': 'odometer_km'}, inplace = True)

autos.head()

date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offer_type   price   abtest vehicle_type  registration_year  \
0  privat    Angebot  5000.0  control          bus               2004   
1  privat    Angebot  8500.0  control    limousine               1997   
2  privat    Angebot  8990.0     test    limousine               2009   
3  privat    Angebot  4350.0  control   kleinwagen               2007   
4  privat    Angebot  1350.0     test        kombi               2003   

     gearbox  power_PS   model  odometer_km  registration_month fuel_type  \
0    manuell       158  andere     150000.0                   3       lpg   
1  automatik       286     7er     150000.0                   6    benzin   
2    manuell       102    golf      70000.0                   7    benzin   
3  automatik        71  fortwo      70000.0                   6    benzin   
4    manuell         0   focus     150000.0                   7    benzin   

        brand unrepaired_damage           ad_created  nr_of_pictures  \
0     peugeot              nein  2016-03-26 00:00:00               0   
1         bmw              nein  2016-04-04 00:00:00               0   
2  volkswagen              nein  2016-03-26 00:00:00               0   
3       smart              nein  2016-03-12 00:00:00               0   
4        ford              nein  2016-04-01 00:00:00               0   

   postal_code            last_seen  
0        79588  2016-04-06 06:45:54  
1        71034  2016-04-06 14:45:08  
2        35394  2016-04-06 20:15:37  
3        33729  2016-03-15 03:16:28  
4        39218  2016-04-01 14:38:50

In [418]:
print(autos['price'].unique().shape)
autos['price'].describe()

(2357,)


count    5.000000e+04
mean     9.840044e+03
std      4.811044e+05
min      0.000000e+00
25%      1.100000e+03
50%      2.950000e+03
75%      7.200000e+03
max      1.000000e+08
Name: price, dtype: float64

In [470]:
autos['price'].value_counts().sort_index(ascending = False).head(20)

99900.0    2
99000.0    2
98500.0    1
94999.0    1
93911.0    1
93000.0    2
89900.0    1
89000.0    1
88900.0    1
86500.0    1
85000.0    1
84997.0    1
84000.0    1
83000.0    1
82987.0    1
80000.0    3
79999.0    1
79980.0    1
79933.0    1
79500.0    1
Name: price, dtype: int64

In [420]:
autos.loc[autos['price'].between(0, 1e5), 'price'].describe()

count    49947.000000
mean      5591.827157
std       7468.390182
min          0.000000
25%       1100.000000
50%       2900.000000
75%       7200.000000
max      99900.000000
Name: price, dtype: float64

In [421]:
print(autos['odometer_km'].unique().shape)
autos['odometer_km'].describe()

(13,)


count     50000.000000
mean     125732.700000
std       40042.211706
min        5000.000000
25%      125000.000000
50%      150000.000000
75%      150000.000000
max      150000.000000
Name: odometer_km, dtype: float64

In [422]:
autos['odometer_km'].value_counts()

150000.0    32424
125000.0     5170
100000.0     2169
90000.0      1757
80000.0      1436
70000.0      1230
60000.0      1164
50000.0      1027
5000.0        967
40000.0       819
30000.0       789
20000.0       784
10000.0       264
Name: odometer_km, dtype: int64

In [423]:
autos = autos[autos['price'].between(0, 1e5)]    # Filter out the outliers

We conclude that there are no noticeable outliers in the data for odometer readings. There were, however, outliers in the data for the price. There were a number of ads with cars listed at \$0, which may have been a consequence of the advertiser neglecting to list the price, intentionally or unintentionally. In addition, there were ads with cars listed at exorbitant prices, the highest being \$99,999,999.00. It is noted that two of the ads had cars listed with the price tag of \$12,345,678.00, a sequence of numbers that is most definitely a spurious number rather than an actual asking price. We established an upper bound of \$10,000, considering many of the prices were in the thousands range (based on the reported values for the mean and quartiles). By establishing this upper bound and neglecting all ads with cars listed under \$0, we have only lost 53 ads out of the original 50,000.

We can work with the remaining datapoints by working with:

autos = autos[autos['prices'].between(0, 1e5)]

which filters out the outliers. We can extract values such as timestamps by calling columns from this filtered dataset:

autos['date_crawled']

From, for example, 'date_crawled', we can extract the year, year/month, or year/month/day by converting the string to numbers, noting that the '-' character separates the numbers. This is assuming that the entries for the date are all formatted in the same fashion. 

# Analysis of Dates

In [424]:
autos[['date_crawled', 'ad_created', 'last_seen']].head()

date_crawled           ad_created            last_seen
0  2016-03-26 17:47:46  2016-03-26 00:00:00  2016-04-06 06:45:54
1  2016-04-04 13:38:56  2016-04-04 00:00:00  2016-04-06 14:45:08
2  2016-03-26 18:57:24  2016-03-26 00:00:00  2016-04-06 20:15:37
3  2016-03-12 16:58:10  2016-03-12 00:00:00  2016-03-15 03:16:28
4  2016-04-01 14:38:50  2016-04-01 00:00:00  2016-04-01 14:38:50

Three of the columns describing date values are string columns. The first ten characters in each column represent the day (e.g. 2016-03-12). 

In [425]:
# To include missing values in the distribution and to use percentages
# instead of counts, chain value_counts(normalize = True, dropna = False)

print(autos['date_crawled'].str[:10].value_counts(normalize = True, dropna = False).sort_index())

print(autos['date_crawled'].str[:10].value_counts(normalize = True, dropna = False).shape)

2016-03-05    0.025367
2016-03-06    0.013955
2016-03-07    0.035998
2016-03-08    0.033255
2016-03-09    0.033215
2016-03-10    0.032154
2016-03-11    0.032514
2016-03-12    0.036779
2016-03-13    0.015577
2016-03-14    0.036659
2016-03-15    0.033976
2016-03-16    0.029511
2016-03-17    0.031513
2016-03-18    0.013034
2016-03-19    0.034937
2016-03-20    0.037840
2016-03-21    0.037480
2016-03-22    0.032835
2016-03-23    0.032394
2016-03-24    0.029111
2016-03-25    0.031754
2016-03-26    0.032474
2016-03-27    0.031033
2016-03-28    0.034837
2016-03-29    0.034176
2016-03-30    0.033596
2016-03-31    0.031914
2016-04-01    0.033776
2016-04-02    0.035438
2016-04-03    0.038681
2016-04-04    0.036519
2016-04-05    0.013094
2016-04-06    0.003183
2016-04-07    0.001422
Name: date_crawled, dtype: float64
(34,)


The ads were added by the crawler from March 5, 2016 to April 7, 2016. Most of the ads were added in March.

In [426]:
print(autos['ad_created'].str[:10].value_counts(normalize = True, dropna = False).sort_index())

print(autos['ad_created'].str[:10].value_counts(normalize = True, dropna = False).shape)

2015-06-11    0.000020
2015-08-10    0.000020
2015-09-09    0.000020
2015-11-10    0.000020
2015-12-05    0.000020
2015-12-30    0.000020
2016-01-03    0.000020
2016-01-07    0.000020
2016-01-10    0.000040
2016-01-13    0.000020
2016-01-14    0.000020
2016-01-16    0.000020
2016-01-22    0.000020
2016-01-27    0.000060
2016-01-29    0.000020
2016-02-01    0.000020
2016-02-02    0.000040
2016-02-05    0.000040
2016-02-07    0.000020
2016-02-08    0.000020
2016-02-09    0.000040
2016-02-11    0.000020
2016-02-12    0.000060
2016-02-14    0.000040
2016-02-16    0.000020
2016-02-17    0.000020
2016-02-18    0.000040
2016-02-19    0.000060
2016-02-20    0.000040
2016-02-21    0.000060
                ...   
2016-03-09    0.033255
2016-03-10    0.031894
2016-03-11    0.032815
2016-03-12    0.036619
2016-03-13    0.016938
2016-03-14    0.035257
2016-03-15    0.033736
2016-03-16    0.030012
2016-03-17    0.031173
2016-03-18    0.013715
2016-03-19    0.033876
2016-03-20    0.037880
2016-03-21 

The ads were created starting from June of 2015. However, the ads from before March 5, 2016 only make up a very small percentage of the ads in this dataset (roughly 0.2%).

In [427]:
print(autos['last_seen'].str[:10].value_counts(normalize = True, dropna = False).sort_index())

print(autos['last_seen'].str[:10].value_counts(normalize = True, dropna = False).shape)

2016-03-05    0.001081
2016-03-06    0.004425
2016-03-07    0.005366
2016-03-08    0.007588
2016-03-09    0.009830
2016-03-10    0.010771
2016-03-11    0.012533
2016-03-12    0.023825
2016-03-13    0.008990
2016-03-14    0.012814
2016-03-15    0.015897
2016-03-16    0.016437
2016-03-17    0.027950
2016-03-18    0.007408
2016-03-19    0.015757
2016-03-20    0.020702
2016-03-21    0.020722
2016-03-22    0.021603
2016-03-23    0.018580
2016-03-24    0.019541
2016-03-25    0.019220
2016-03-26    0.016918
2016-03-27    0.016017
2016-03-28    0.020842
2016-03-29    0.022344
2016-03-30    0.024866
2016-03-31    0.023845
2016-04-01    0.023125
2016-04-02    0.024886
2016-04-03    0.025367
2016-04-04    0.024646
2016-04-05    0.124152
2016-04-06    0.221014
2016-04-07    0.130939
Name: last_seen, dtype: float64
(34,)


A large number of the ads were last seen from April 5, 2016 to April 7, 2016. Only a small proportion of the ads were last seen in early March.

## Registration Year

In [428]:
autos['registration_year'].describe()

count    49947.0000
mean      2005.0755
std        105.7677
min       1000.0000
25%       1999.0000
50%       2003.0000
75%       2008.0000
max       9999.0000
Name: registration_year, dtype: float64

The reported mean registration year is 2005. However, the registration year data has outliers: The minimum year reported is 1000 (well before the invention of cars) and the maximum year reported is 9999 (currently 7,981 years into the future (7,983 years into the future from when this dataset was compiled)). The values for the 25%, 50%, and 75% percentiles currently provide a more accurate depiction of the spread of registration year values.

In [429]:
autos.loc[((autos['registration_year'] < 1910) | (autos['registration_year'] > 2016)), 'registration_year'].value_counts().sort_index()

1000       1
1001       1
1111       1
1500       1
1800       2
2017    1452
2018     491
2019       3
2800       1
4100       1
4500       1
4800       1
5000       4
5911       1
6200       1
8888       1
9000       2
9996       1
9999       4
Name: registration_year, dtype: int64

The lowest acceptable value for registration_year is, as suggested, 1900. The only reported years before 1900 are 1000, 1001, 1111, 1500, and 1800, all before the car was invented. We have decided that the highest acceptable year for registration year will be 2017, as people may have registered their vehicles for the next year early in 2016. This accounts for why 1452 data points have the registration year of 2017.

In [430]:
autos = autos[(autos['registration_year'] >= 1900) & (autos['registration_year'] <= 2017)]

In [431]:
print(autos['registration_year'].value_counts(normalize = True).sort_index(ascending = False))
print(autos['registration_year'].describe())

2017    0.029375
2016    0.026503
2015    0.007991
2014    0.013413
2013    0.016225
2012    0.026705
2011    0.033037
2010    0.032228
2009    0.042424
2008    0.045095
2007    0.046592
2006    0.054765
2005    0.060997
2004    0.055372
2003    0.055170
2002    0.051245
2001    0.054644
2000    0.067835
1999    0.060653
1998    0.049566
1997    0.041008
1996    0.029214
1995    0.026523
1994    0.013352
1993    0.009003
1992    0.007870
1991    0.007202
1990    0.007991
1989    0.003662
1988    0.002853
          ...   
1968    0.000506
1967    0.000546
1966    0.000445
1965    0.000344
1964    0.000243
1963    0.000182
1962    0.000081
1961    0.000121
1960    0.000668
1959    0.000142
1958    0.000081
1957    0.000040
1956    0.000101
1955    0.000040
1954    0.000040
1953    0.000020
1952    0.000020
1951    0.000020
1950    0.000061
1948    0.000020
1943    0.000020
1941    0.000040
1939    0.000020
1938    0.000020
1937    0.000081
1934    0.000040
1931    0.000020
1929    0.0000

Most of the vehicles listed in the dataset were registered from 1990 onward, and the greatest fraction of the vehicles were registered from 1999 to 2010. Only a tiny fraction of vehicles were registered from 1910 (the earliest year) the 60s, 70s, or 80s. Once again, the values reported for the 25%, 50%, and 75% percentiles accurately depict the spread of reported years. It is seen that the mean year is now 2003.

# Analysis of Car Brands

Let us look at the car brands. This analysis will involve aggregation.

In [432]:
autos['brand'].unique()

print(len(autos['brand'].unique()))

40


However, there are very many brands. Let us select those that have over 1% of the total values.

In [433]:
top_brands = autos['brand'].value_counts(ascending = False, normalize = True) > 0.05
top_brands = top_brands[top_brands[top_brands.index] == True]

print(top_brands)
print(top_brands.index.shape)

volkswagen       True
opel             True
bmw              True
mercedes_benz    True
audi             True
ford             True
Name: brand, dtype: bool
(6,)


There are 40 brands of cars in the dataset of approximately 50,000 cars (outlier ads were previously filtered out). We are going to look at the car brands that make up over 5% of the total values. This filter has reduced the number of brands to 6. It is likely that this filter gives us a look at which car brands are the most popular in Germany. It is not surprising that most of these top brands are German car brands.

In [434]:
brand_mean_prices = {}

for brand in top_brands.index:
    brand_mean_prices[brand] = autos.loc[autos['brand'] == brand, 'price'].mean()
    
print(brand_mean_prices)
print('Brand with minimum mean price: ' + min(brand_mean_prices, key = brand_mean_prices.get))
print('Brand with maximum mean price: ' + max(brand_mean_prices, key = brand_mean_prices.get))

{'volkswagen': 5177.634857521537, 'opel': 2853.8153988868276, 'bmw': 7926.173452314557, 'mercedes_benz': 8289.653649167734, 'audi': 8950.519162943805, 'ford': 3596.0690255220416}
Brand with minimum mean price: opel
Brand with maximum mean price: audi


We see that some of the most expensive brands in this dataset are Mercedes Benz, BMW, and Audi. Some of the least expensive brands in this filtered dataset are Opel and Ford (the only non-German car brand here, largest US car brand). Volkswagen falls in between.

In [435]:
mean_mileage = {}

for brand in top_brands.index:
    mean_mileage[brand] = autos.loc[autos['brand'] == brand, 'odometer_km'].mean()
    
print(mean_mileage)

{'volkswagen': 128970.46293666572, 'opel': 129322.82003710575, 'bmw': 132575.75757575757, 'mercedes_benz': 131097.9513444302, 'audi': 129590.87702798026, 'ford': 124100.92807424594}


In [436]:
# Convert both dictionaries to Series
bmp_series = pd.Series(brand_mean_prices)
mile_series = pd.Series(mean_mileage)

brand_df = pd.DataFrame(bmp_series, columns = ['mean_prices'])
brand_df['mean_mileage'] = mile_series
brand_df

mean_prices   mean_mileage
volkswagen     5177.634858  128970.462937
opel           2853.815399  129322.820037
bmw            7926.173452  132575.757576
mercedes_benz  8289.653649  131097.951344
audi           8950.519163  129590.877028
ford           3596.069026  124100.928074

Looking at the top brands in the dataset (over 5% of the total), we see that the brand with the lowest mean price, Opel, has an in-between mean mileage, and the brand with the highest mean price, Audi, also has an in-between mean mileage (very similar to that of Opel). The brand with the lowest mean mileage, Ford, has the second lowest mean price, and the brand with the highest mean mileage, BMW, has the third highest mean price. From these six brands alone, it is hard to find a definite trend between the mean prices and mean mileages of the brands.

Next steps to consider:

- Data cleaning next steps:
 - Identify categorical data that uses German words, translate them, and map the values to their English counterparts
 - Convert the dates to be uniform numeric data, so "2016-03-21" becomes the integer 20160321.
 - See if there are particular keywords in the name column that you can extract as new columns.

- Analysis next steps:
 - Find the most common brand/model combinations.
 - Split the odometer_km into groups, and use aggregation to see if average prices follow any patterns based on the mileage.
 - How much cheaper are cars with damage than their non-damaged counterparts?

# Continuing Past DataQuest Mission - Further Data Cleaning

## Categorical Data - German --> English

In [437]:
autos.head()

date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offer_type   price   abtest vehicle_type  registration_year  \
0  privat    Angebot  5000.0  control          bus               2004   
1  privat    Angebot  8500.0  control    limousine               1997   
2  privat    Angebot  8990.0     test    limousine               2009   
3  privat    Angebot  4350.0  control   kleinwagen               2007   
4  privat    Angebot  1350.0     test        kombi               2003   

     gearbox  power_PS   model  odometer_km  registration_month fuel_type  \
0    manuell       158  andere     150000.0                   3       lpg   
1  automatik       286     7er     150000.0                   6    benzin   
2    manuell       102    golf      70000.0                   7    benzin   
3  automatik        71  fortwo      70000.0                   6    benzin   
4    manuell         0   focus     150000.0                   7    benzin   

        brand unrepaired_damage           ad_created  nr_of_pictures  \
0     peugeot              nein  2016-03-26 00:00:00               0   
1         bmw              nein  2016-04-04 00:00:00               0   
2  volkswagen              nein  2016-03-26 00:00:00               0   
3       smart              nein  2016-03-12 00:00:00               0   
4        ford              nein  2016-04-01 00:00:00               0   

   postal_code            last_seen  
0        79588  2016-04-06 06:45:54  
1        71034  2016-04-06 14:45:08  
2        35394  2016-04-06 20:15:37  
3        33729  2016-03-15 03:16:28  
4        39218  2016-04-01 14:38:50

In [438]:
print(autos['seller'].unique())
print(autos['vehicle_type'].unique())
print(autos['gearbox'].unique())
print(autos['fuel_type'].unique())
print(autos['unrepaired_damage'].unique())

['privat' 'gewerblich']
['bus' 'limousine' 'kleinwagen' 'kombi' nan 'coupe' 'suv' 'cabrio'
 'andere']
['manuell' 'automatik' nan]
['lpg' 'benzin' 'diesel' nan 'cng' 'hybrid' 'elektro' 'andere']
['nein' nan 'ja']


Let us look at the seller, vehicle_type, gearbox, fuel_type, and unrepaired_damage columns, which contain categorical data. Considering this is data pulled from German eBay, the data (strings) are written in German. Let us try to translate them to English (emphasis on "try" due to a lack of fluency in German by this data analyst).

In [439]:
seller_translate = {'privat': 'private', 'gewerblich': 'commercial'}
vehicle_translate = {'bus': 'bus', 'limousine': 'limousine', 'kleinwagen': 'small car', 'kombi': 'station wagon', 
                     'coupe': 'coupe', 'suv': 'suv', 'cabrio': 'convertible', 'andere': 'other'}   # Cabrio is English. Also, convertible
gearbox_translate = {'manuell': 'manual', 'automatik': 'automatic'}
fuel_translate = {'lpg': 'lpg', 'benzin': 'petrol', 'diesel': 'diesel', 'cng': 'cng', 'hybrid': 'hybrid',
                  'elektro': 'electric', 'andere': 'other'}
damage_translate = {'nein': 'no', 'ja': 'yes'}

In [440]:
autos['seller'] = autos['seller'].map(seller_translate)
autos['vehicle_type'] = autos['vehicle_type'].map(vehicle_translate)
autos['gearbox'] = autos['gearbox'].map(gearbox_translate)
autos['fuel_type'] = autos['fuel_type'].map(fuel_translate)
autos['unrepaired_damage'] = autos['unrepaired_damage'].map(damage_translate)

In [441]:
autos.head()

date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

    seller offer_type   price   abtest   vehicle_type  registration_year  \
0  private    Angebot  5000.0  control            bus               2004   
1  private    Angebot  8500.0  control      limousine               1997   
2  private    Angebot  8990.0     test      limousine               2009   
3  private    Angebot  4350.0  control      small car               2007   
4  private    Angebot  1350.0     test  station wagon               2003   

     gearbox  power_PS   model  odometer_km  registration_month fuel_type  \
0     manual       158  andere     150000.0                   3       lpg   
1  automatic       286     7er     150000.0                   6    petrol   
2     manual       102    golf      70000.0                   7    petrol   
3  automatic        71  fortwo      70000.0                   6    petrol   
4     manual         0   focus     150000.0                   7    petrol   

        brand unrepaired_damage           ad_created  nr_of_pictures  \
0     peugeot                no  2016-03-26 00:00:00               0   
1         bmw                no  2016-04-04 00:00:00               0   
2  volkswagen                no  2016-03-26 00:00:00               0   
3       smart                no  2016-03-12 00:00:00               0   
4        ford                no  2016-04-01 00:00:00               0   

   postal_code            last_seen  
0        79588  2016-04-06 06:45:54  
1        71034  2016-04-06 14:45:08  
2        35394  2016-04-06 20:15:37  
3        33729  2016-03-15 03:16:28  
4        39218  2016-04-01 14:38:50

## Convert Dates to Integers

The data to convert are in the columns data_crawled, ad_created, and last_seen. We will be dropping the time, so we will be looking at the first ten characters.

In [442]:
for col in ['date_crawled', 'ad_created', 'last_seen']:
    autos[col] = autos[col].str[:10]
    autos[col] = autos[col].str.replace('-','').astype(int)
autos.head()

date_crawled                                               name   seller  \
0      20160326                   Peugeot_807_160_NAVTECH_ON_BOARD  private   
1      20160404         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik  private   
2      20160326                         Volkswagen_Golf_1.6_United  private   
3      20160312  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...  private   
4      20160401  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...  private   

  offer_type   price   abtest   vehicle_type  registration_year    gearbox  \
0    Angebot  5000.0  control            bus               2004     manual   
1    Angebot  8500.0  control      limousine               1997  automatic   
2    Angebot  8990.0     test      limousine               2009     manual   
3    Angebot  4350.0  control      small car               2007  automatic   
4    Angebot  1350.0     test  station wagon               2003     manual   

   power_PS   model  odometer_km  registration_month fuel_type       brand  \
0       158  andere     150000.0                   3       lpg     peugeot   
1       286     7er     150000.0                   6    petrol         bmw   
2       102    golf      70000.0                   7    petrol  volkswagen   
3        71  fortwo      70000.0                   6    petrol       smart   
4         0   focus     150000.0                   7    petrol        ford   

  unrepaired_damage  ad_created  nr_of_pictures  postal_code  last_seen  
0                no    20160326               0        79588   20160406  
1                no    20160404               0        71034   20160406  
2                no    20160326               0        35394   20160406  
3                no    20160312               0        33729   20160315  
4                no    20160401               0        39218   20160401

## Extracting Keywords from "name"

In [471]:
#autos['name'] = autos['name'].str.replace('Mercedes_Benz', 'Mercedes')
#autos['name'] = autos['name'].str.replace('vw_', 'volkswagen_')
#autos['name'] = autos['name'].str.replace('w.', 'volkswagen_')
#autos['name'] = autos['name'].str.lower()
autos['name'].str.rsplit('_', expand = True, n = 2).head()

0             1  \
0                 Peugeot_807_160_NAVTECH            ON   
1         BMW_740i_4_4_Liter_HAMANN_UMBAU          Mega   
2                         Volkswagen_Golf           1.6   
3                      Smart_smart_fortwo         coupe   
4  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr  gepflegt.mit   

                            2  
0                       BOARD  
1                       Optik  
2                      United  
3  softouch/F1/Klima/Panorama  
4                 Klimaanlage

An issue from extracting words from the name column is that for most of the data entires, the first two words are the brand and model, which are already existing columns. It may be argued that the rest of the words are not optimal for categorization purposes, however.

The lines of code can be written to clean up the name column and extract the brand and model names, which can be written into separate columns of a DataFrame. In this case, they already exist:

autos['name'] = autos['name'].str.replace('Mercedes_Benz', 'Mercedes')

autos['name'] = autos['name'].str.replace('vw_', 'volkswagen_')

autos['name'] = autos['name'].str.lower()

# Further Analysis

## Most Common Brand/Model Combinations

In [460]:
# May want to look within top brands and top models.

for brand in top_brands.index:
    brand_auto = autos[autos['brand'] == brand]
    top_models = brand_auto['model'].value_counts(ascending = False, normalize = True) > 0.1
    #print(brand)
    #print(top_models[top_models[top_models.index] == True].index[:3])
    
    model_series = pd.Series(top_models[top_models[top_models.index] == True].index[:3])
    model_df = pd.DataFrame(model_series, columns = ['Brand = ' + brand])
    
    print(model_df)

  Brand = volkswagen
0               golf
1               polo
2             passat
  Brand = opel
0        corsa
1        astra
2       vectra
  Brand = bmw
0         3er
1         5er
2         1er
  Brand = mercedes_benz
0              c_klasse
1              e_klasse
2              a_klasse
  Brand = audi
0           a4
1           a3
2           a6
  Brand = ford
0        focus
1       fiesta
2       mondeo


## Grouping odometer_km

In [462]:
autos['odometer_km'].value_counts()

150000.0    32069
125000.0     5120
100000.0     2145
90000.0      1738
80000.0      1420
70000.0      1220
60000.0      1156
50000.0      1017
5000.0        930
40000.0       814
30000.0       778
20000.0       770
10000.0       252
Name: odometer_km, dtype: int64

In [464]:
odometer_high = autos[autos['odometer_km'] >= 100000]
odometer_mid = autos[(autos['odometer_km'] >= 50000) & (autos['odometer_km'] < 100000)]
odometer_low = autos[autos['odometer_km'] < 50000]

In [466]:
print(odometer_high['price'].mean())
print(odometer_mid['price'].mean())
print(odometer_low['price'].mean())

4138.323867392078
10412.711494428331
13062.693002257336


The average price goes down with greater mileage. 

## Relation Between Cost and Damage

In [467]:
autos['unrepaired_damage'].unique()

array(['no', nan, 'yes'], dtype=object)

We will be looking at the average price of the cars with damage ('yes') and without damage ('no'). We will not be counting cars for which no answer is reported (nan).

In [468]:
print(autos.loc[autos['unrepaired_damage'] == 'no', 'price'].mean())
print(autos.loc[autos['unrepaired_damage'] == 'yes', 'price'].mean())

6845.638634996132
2110.092110614071


On average, cars with damage are cheaper than their non-damaged counterparts by $4700. We can continue this analysis by doing this comparison for each top brand.

In [469]:
for brand in top_brands.index:
    brand_auto = autos[autos['brand'] == brand]
    undamaged = brand_auto.loc[brand_auto['unrepaired_damage'] == 'no', 'price'].mean()
    damaged = brand_auto.loc[brand_auto['unrepaired_damage'] == 'yes', 'price'].mean()
    
    damage_series = pd.Series([undamaged, damaged], index = ['Undamaged', 'Damaged'])
    damage_df = pd.DataFrame(damage_series, columns = [brand])
    print(damage_df)

            volkswagen
Undamaged  6329.825412
Damaged    2018.663090
                  opel
Undamaged  3585.791361
Damaged    1291.547731
                   bmw
Undamaged  9179.448506
Damaged    3286.546067
           mercedes_benz
Undamaged    9511.595594
Damaged      3807.539326
                   audi
Undamaged  10656.891883
Damaged     3140.718663
                  ford
Undamaged  4600.667854
Damaged    1299.495556
